crates/wasmparser/src/validator/operators.rs

```diff
@@ -1416,6 +1428,65 @@ where
         }
     }
 
+    fn cont_type_at(&self, key: TypeKey) -> Result<&ContType> {
+        let sub_ty = match key {
+            Either::A(at) => self.sub_type_at(at)?,
+            Either::B(id) => match self.resources.sub_type_at_id(id) {
+                None => bail!(self.offset, "unknown type {}", TypeIdentifier::index(&id)),
+                Some(sub_ty) => sub_ty,
+            },
+        };
+        if let CompositeInnerType::Cont(cont_ty) = &sub_ty.composite_type.inner {
+            if self.inner.shared && !sub_ty.composite_type.shared {
+                bail!(
+                    self.offset,
+                    "shared continuations cannot access unshared continuations",
+                );
+            }
+            Ok(cont_ty)
+        } else {
+            bail!(
+                self.offset,
+                "non-continuation type {}",
+                match key {
+                    Either::A(at) => at as usize,
+                    Either::B(id) => TypeIdentifier::index(&id),
+                }
+            )
+        }
+    }
+
+    fn func_type_of_cont_type_at(&self, key: TypeKey) -> Result<&'resources FuncType> {
+        let cont_ty = self.cont_type_at(key)?;
+        match cont_ty.0.as_core_type_id() {
+            None => bail!(self.offset, "invalid cont type"),
+            Some(func_id) => match self.resources.sub_type_at_id(func_id) {
+                None => bail!(
+                    self.offset,
+                    "unknown type {}",
+                    TypeIdentifier::index(&func_id)
+                ),
+                Some(sub_ty) => {
+                    if let CompositeInnerType::Func(func_ty) = &sub_ty.composite_type.inner {
+                        if self.inner.shared && !sub_ty.composite_type.shared {
+                            bail!(
+                                self.offset,
+                                "shared functions cannot access unshared functions",
+                            );
+                        }
+                        Ok(&func_ty)
+                    } else {
+                        bail!(
+                            self.offset,
+                            "non-function type {}",
+                            TypeIdentifier::index(&func_id)
+                        )
+                    }
+                }
+            },
+        }
+    }
```

Comment:
This function, if it takes `&ContType`, I believe can be infallible (e.g. unwraps/expects/etc). The validation that `ContType` always points to a function should have already been validated when the type was added to the module, so here's it's safe to assume that invariant is always upheld.

Reply:
Great suggestion, thanks! After changing the index representation on `Cont` to `PackedIndex` everything else kind of falls into place. It is much nicer.